sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/TableChange.java

```diff
@@ -0,0 +1,293 @@
+   */
+  static TableChange addColumn(String[] fieldNames, DataType dataType) {
+    return new AddColumn(fieldNames, dataType, true, null);
+  }
+
+  /**
+   * Create a TableChange for adding a column to a table.
+   *
+   * @param fieldNames field names of the new column
+   * @param dataType the new column's data type
+   * @param isNullable whether the new column can contain null
+   * @return a TableChange for the addition
+   */
+  static TableChange addColumn(String[] fieldNames, DataType dataType, boolean isNullable) {
+    return new AddColumn(fieldNames, dataType, isNullable, null);
+  }
+
+  /**
+   * Create a TableChange for adding a top-level column to a table.
+   *
+   * @param fieldNames field names of the new column
+   * @param dataType the new column's data type
+   * @param isNullable whether the new column can contain null
+   * @param comment the new field's comment string
+   * @return a TableChange for the addition
+   */
+  static TableChange addColumn(
+      String[] fieldNames,
+      DataType dataType,
+      boolean isNullable,
+      String comment) {
+    return new AddColumn(fieldNames, dataType, isNullable, comment);
+  }
+
+  /**
+   * Create a TableChange for renaming a field.
+   * <p>
+   * The name is used to find the field to rename. The new name will replace the leaf field name.
+   * For example, renameColumn("a.b.c", "x") should produce column a.b.x.
+   *
+   * @param fieldNames the current field names
+   * @param newName the new name
+   * @return a TableChange for the rename
+   */
+  static TableChange renameColumn(String[] fieldNames, String newName) {
+    return new RenameColumn(fieldNames, newName);
+  }
+
+  /**
+   * Create a TableChange for updating the type of a field.
+   * <p>
+   * The field names are used to find the field to update.
+   *
+   * @param fieldNames field names of the column to update
+   * @param newDataType the new data type
+   * @return a TableChange for the update
+   */
+  static TableChange updateColumn(String[] fieldNames, DataType newDataType) {
+    return new UpdateColumn(fieldNames, newDataType, true);
+  }
+
+  /**
+   * Create a TableChange for updating the type of a field.
+   * <p>
+   * The field names are used to find the field to update.
+   *
+   * @param fieldNames field names of the column to update
+   * @param newDataType the new data type
+   * @return a TableChange for the update
+   */
+  static TableChange updateColumn(String[] fieldNames, DataType newDataType, boolean isNullable) {
+    return new UpdateColumn(fieldNames, newDataType, isNullable);
+  }
+
+  /**
+   * Create a TableChange for updating the comment of a field.
+   * <p>
+   * The name is used to find the field to update.
+   *
+   * @param fieldNames field names of the column to update
+   * @param newComment the new comment
+   * @return a TableChange for the update
+   */
+  static TableChange updateComment(String[] fieldNames, String newComment) {
```

Comment:
We don't want to pass data from the source back into the source. For example, if a user runs `ALTER TABLE t ALTER COLUMN c COMMENT 'some doc'`, why would we pass the type of column `c` back to the source? If Spark passed the type, then the source would need to detect what changed (the comment) and what didn't (the type) even though Spark already had that information.

It is also not correct to pass the type in when the user only requested a change to the comment. If the table has changed since the source loaded it, passing the type could tell the source to revert another change. Regardless of whether this is likely to happen, it would be a correctness bug.

Reply:
Sounds reasonable. Shall we do it to all the column fields? like changing column data type only, changing column nullable only. Then we don't need the general UpdateColumn.

Reply:
@cloud-fan, no. The type of a field is both the `DataType` and nullability in SQL, which is why types are expressed as `data_type [NOT NULL]`.

For example:
```
ALTER TABLE t ALTER COLUMN c COMMENT 'doc'
ALTER TABLE t ALTER COLUMN i TYPE bigint
ALTER TABLE t ALTER COLUMN f TYPE double NOT NULL
ALTER TABLE t ALTER COLUMN d TYPE decimal(18,2) NOT NULL COMMENT 'doc'
```

Reply:
To confirm I understand your statement, are you saying
1. data type and nullable together mean the field type, although users can alter one of them via SQL command, the catalog should always update them together.
2. comment is a separated property and can be updated separately.

If this is true, I think we should have 3 methods: `updateType`, `updateComment` and `updateColumn`.

Reply:
We want to keep it simple and avoid creating combinations of changes. If a user makes multiple changes at the same time, they are all passed to `alterTable`. That avoids the need to create combinations of these changes, like `updateColumn` that combines updating the comment and updating the type.